src/oc/auth/async/expo.clj

```diff
@@ -0,0 +1,163 @@
+        bad-receipt-ids               (keep bad-receipt-kv receipt-tuples)
+        bad-push-tokens-from-receipts (->> bad-receipt-ids
+                                           (mapcat (comp #(map :pushToken %) by-ticket-id))
+                                           (into #{}))]
+    (into bad-tokens-from-tickets
+          bad-push-tokens-from-receipts)))
+
+(defn- remove-bad-push-tokens-from-users!
+  [db-pool push-notifications bad-push-tokens]
+  (let [bad-push-notifications (filter (comp bad-push-tokens :pushToken) push-notifications)
+        push-notif-user-id     (comp :user-id :data)
+        push-token->user-id    (zipmap (map :pushToken bad-push-notifications)
+                                       (map push-notif-user-id bad-push-notifications))]
+    (pool/with-pool [conn db-pool]
+      (doseq [[push-token user-id] push-token->user-id]
+        (let [user            (user-res/get-user conn user-id)
+              old-push-tokens (into #{} (:expo-push-tokens user))
+              new-push-tokens (seq (disj old-push-tokens push-token))]
+          (timbre/info "Removing bad push tokens from user" {:user-id user-id
+                                                             :push-token push-token
+                                                             :old-push-tokens old-push-tokens
+                                                             :new-push-tokens new-push-tokens})
+          (user-res/update-user! conn user-id {:expo-push-tokens new-push-tokens}))))))
+
+(defn- handle-expo-message
+  [db-pool msg]
+  (let [{:keys [push-notifications tickets]} msg]
+    (when (and (seq push-notifications) (seq tickets))
+      (let [receipts (get-receipts tickets)
+            bad-push-tokens (determine-bad-push-tokens push-notifications tickets receipts)]
+        (timbre/info "Checked for bad push tokens" (merge msg {:receipts receipts}))
+        (when (not-empty bad-push-tokens)
+          (try
+            (remove-bad-push-tokens-from-users! db-pool push-notifications bad-push-tokens)
+            (catch Exception e
+              (throw (ex-info (.getMessage e)
+                              (merge msg {:receipts receipts
+                                          :bad-push-tokens bad-push-tokens})))
+              )))))))
+
+(comment
+
+  ;; Sample receipts
+  ;; [{:06c938b1-aca5-47fc-8750-d23ea257bae8 {:status "error"}}]
+
+  (let [{:keys [push-notifications tickets] :as msg}
+        {:push-notifications [{:pushToken "TOKEN_A"}
+                              {:pushToken "TOKEN_B"}
+                              {:pushToken "TOKEN_C"}]
+         :tickets            [{:status "ok" :id "ID_A"}
+                              {:status "ok" :id "ID_B"}
+                              {:status "error" :id "ID_C"}]}
+        receipts [{:ID_A {:status "ok"}
+                   :ID_B {:status "error"}}
+                  {:ID_C {:status "ok"}}]]
+    (determine-bad-push-tokens push-notifications tickets receipts)
```

Comment:
@cjsauer maybe it's only this code that is wrong but running this from the repl i get this map as result of `determine-bad-push-tokens` and as you can see the user-id are nil:
`{"TOKEN_B" nil, "TOKEN_C" nil}`
Maybe i did something wrong or maybe the starting map here is wrong.
I am starting now to do the test steps so i'll be testing the token remove on Android soon.